BrowserKit/Sources/TabDataStore/TabDataStore.swift

```diff
@@ -3,19 +3,208 @@
+        }
+        return nil
+    }
+
+    //MARK: Fetching Window Data
+    func fetchWindowData() async -> WindowData {
         return WindowData(id: UUID(), isPrimary: true, activeTabId: UUID(), tabData: [])
     }
 
-    func saveTabData(window: WindowData) async {}
+    func fetchWindowData(withId id: UUID) async -> WindowData? {
+        guard let profileURL = windowDataDirectoryURL else {
+            return nil
+        }
+
+        do {
+            let fileURLs = try FileManager.default.contentsOfDirectory(at: profileURL, includingPropertiesForKeys: nil, options: .skipsHiddenFiles)
+
+            let windowDataFiles = fileURLs.filter { $0.path.contains(DefaultTabDataStore.storePath) }
+
+            for fileURL in windowDataFiles {
+                do {
+                    let data = try Data(contentsOf: fileURL)
+                    let windowData = try JSONDecoder().decode(WindowData.self, from: data)
+                    if windowData.id == id {
+                        return windowData
+                    }
+                } catch {
+                    logger.log("Error decoding window data: \(error)",
+                               level: .debug,
+                               category: .tabs)
+                }
+            }
+        } catch {
+            logger.log("Error fetching window data with ID: \(error)",
+                       level: .debug,
+                       category: .tabs)
+        }
+
+        return nil
+    }
+
+    func fetchAllWindowsData() async -> [WindowData] {
+        guard let profileURL = windowDataDirectoryURL else {
+            return []
+        }
+
+        do {
+            let fileURLs = try FileManager.default.contentsOfDirectory(at: profileURL, includingPropertiesForKeys: nil, options: .skipsHiddenFiles)
+
+            let windowDataFiles = fileURLs.filter { $0.path.contains(DefaultTabDataStore.storePath) }
+
+            var windowsData: [WindowData] = []
+            for fileURL in windowDataFiles {
+                do {
+                    let data = try Data(contentsOf: fileURL)
+                    let windowData = try JSONDecoder().decode(WindowData.self, from: data)
+                    windowsData.append(windowData)
+                } catch {
+                    logger.log("Error decoding window data: \(error)",
+                               level: .debug,
+                               category: .tabs)
+                }
+            }
+            return windowsData
+        } catch {
+            logger.log("Error fetching all window data: \(error)",
+                       level: .debug,
+                       category: .tabs)
+            return []
+        }
+    }
+
+    //MARK: Saving Data
+    func saveWindowData(window: WindowData) async {
+        saveWorkItem?.cancel()
+
+        let workItem = DispatchWorkItem { [weak self] in
+            guard let self = self else { return }
+            Task.init(priority: .userInitiated) {
```

Comment:
I thought that saving and retrieving tabs will happen because the user creates/removes a tab and thus the system should give this task a higher priority than default

Reply:
Since we are throttling the request anyway I don't think it makes sense to force it to run at a higher priority. The issue I'm trying to avoid is running multiple nested tasks where there are conflicting priorities which is an issue we have in the app at the moment. So I prefer to only see the priority defined as anything other than default if there is a strong reason to do so.